[TensorExpr] Add serializer / deserializer

eellison opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#46553 [TensorExpr] Add serializer / deserializer**
* #46552 Make TensorExpr Enums convertable to/from Strings
* #46551 [JIT] [Temp] Add dependency

This PR adds conversion to and from json to NNC. I have taken a dependency on https://github.com/nlohmann/json for json, which I talked to @malfet, the build lead about. 

A NNC Expression like the following:
`y * 2.f + (float(x) * 3.f + 4.f * y)`

Gets serialized as:
```
{
  "dtype": "Float",
  "expr_type": "Add",
  "json_id": 10,
  "lhs": {
    "dtype": "Float",
    "expr_type": "Mul",
    "json_id": 2,
    "lhs": {
      "dtype": "Float",
      "expr_type": "Var",
      "json_id": 0,
      "name": "y"
    },
    "rhs": {
      "dtype": "Float",
      "expr_type": "FloatImm",
      "json_id": 1,
      "value": 2
    }
  },
  "rhs": {
    "dtype": "Float",
    "expr_type": "Add",
    "json_id": 9,
    "lhs": {
      "dtype": "Float",
      "expr_type": "Mul",
      "json_id": 6,
      "lhs": {
        "dtype": "Float",
        "expr_type": "Cast",
        "json_id": 4,
        "src_value": {
          "dtype": "Half",
          "expr_type": "Var",
          "json_id": 3,
          "name": "x"
        }
      },
      "rhs": {
        "dtype": "Float",
        "expr_type": "FloatImm",
        "json_id": 5,
        "value": 3
      }
    },
    "rhs": {
      "dtype": "Float",
      "expr_type": "Mul",
      "json_id": 8,
      "lhs": {
        "dtype": "Float",
        "expr_type": "FloatImm",
        "json_id": 7,
        "value": 4
      },
      "rhs": {
        "dtype": "Float",
        "expr_type": "Var",
        "json_id": 0,
        "name": "y"
      }
    }
  }
```

A few notes about implementation: 

- JSON has no builtin way to resolve references, so serialized KernelScopedObjects are cached, and assigned a json_id. When we deserialize, we look if we already deserialized a node with that ID in order to maintain multiple references to an object. 

- This memoization is currently broken just for Tensors, so each repeated call to a Tensor increments the name once. I've tried debugging it and would appreciate some help here 😛 This may be because there is no visitor defined for Tensors/Functions, just FunctionCall nodes. I'm not sure exactly though...

- `ReduceOp` is not able to be serialized currently because `interaction` is stored as a lambda. I'm not sure why there is the indirection there instead of directly storing an Expression or function or something.

- Testing : this is certainly light on testing currently. In JIT, any time we compile a graph in test files we test that it serializes correctly. It would be nice to do this also for TensorExprs. This is currently blocked by ReduceOp and Tensor memoization issue I mentioned.

Commit:
Add serializer / deserializer

[ghstack-poisoned]

Commit:
Update on "Add serializer / deserializer"




[ghstack-poisoned]

ZolotukhinM commented:
I haven't looked into it deeply yet, but that's a great feature to have!
CCing @bertmaher and @nickgg as well

bertmaher commented:
Cool!!!

nickgg commented:
Real cool. I'll take a look today.

My first instinct is that json is overkill for this since our IR has such a regular structure. I was picturing something like a lisp like format. Doing one doesn't prevent us doing the other if we think we need it, so not a blocker, but if it means adding a new dependency it's work thinking about.

Re Tensors, they're not really part of the IR so I wouldn't include them here (instead would recommend the Var from the relevant Buf).

Re ReduceOp, yeah thats a problem with the design of Reductions that we need to fix as it prevents mutators from working on the interaction (the biggest issue being Var substitution can't guarantee it found all usages of a Var). If you're interested in fixing this that would be awesome and I'd be happy to talk you through it.

nickgg commented on torch/csrc/jit/tensorexpr/ir_deserializer.cpp:
```diff
@@ -0,0 +1,345 @@
+  CHECK_EXPR_TYPE(IfThenElse);
+  // CHECK_EXPR_TYPE(Term);
+  // CHECK_EXPR_TYPE(Polynomial);
+  CHECK_EXPR_TYPE(MaxTerm);
```
MinTerm and MaxTerm have the same lifetime as Term and Polynomial (ie. they exist only during arithmetic simplification) so if you're not including Term and Poly (which sounds right to me) then I wouldn't include these as well.

Min and Max (binary ops) are important though.

nickgg commented on torch/csrc/jit/tensorexpr/ir_deserializer.h:
```diff
@@ -0,0 +1,234 @@
+  const IntImm* deserializeIntImm(const json& json);
+  const ByteImm* deserializeByteImm(const json& json);
+
+  // void visit(const Load* v) override;
```
You are missing an implementation for Load, which seems critical. This would also be a strong candidate to test.

nickgg commented on torch/csrc/jit/tensorexpr/ir_deserializer.h:
```diff
@@ -0,0 +1,234 @@
+  //   AT_FORALL_SCALAR_TYPES_AND2(Bool, Half, IMM_SERIALIZE_VISIT);
+  // #undef IMM_SERIALIZE_VISIT
+
+  //   void visit(const Cast* v) override;
```
Once copy of these comments is definitely enough

eellison replied:
Lol, sorry I thought I delete all of this stuff, maybe I forgot to re-ghstack.

nickgg commented on torch/csrc/jit/tensorexpr/ir_serializer.cpp:
```diff
@@ -0,0 +1,557 @@
+          {"value", v->value()},
+          {"dtype", serializeDtype(v->dtype())},
+      });
+}
+
+void torch::jit::tensorexpr::IRSerializer::visit(
+    torch::jit::tensorexpr::HalfImm const* v) {
+  JSON_CACHE_GUARD();
+  putJSON(
+      v,
+      {
+          {"expr_type", "HalfImm"},
+          {"value", static_cast<float>(v->value())},
+          {"dtype", serializeDtype(v->dtype())},
+      });
+}
+
+void torch::jit::tensorexpr::IRSerializer::visit(
+    torch::jit::tensorexpr::ByteImm const* v) {
+  JSON_CACHE_GUARD();
+  putJSON(
+      v,
+      {
+          {"expr_type", "ByteImm"},
+          {"value", v->value()},
+          {"dtype", serializeDtype(v->dtype())},
+      });
+}
+
+void torch::jit::tensorexpr::IRSerializer::visit(
+    torch::jit::tensorexpr::IntImm const* v) {
+  JSON_CACHE_GUARD();
+  putJSON(
+      v,
+      {
+          {"expr_type", "IntImm"},
+          {"value", v->value()},
+          {"dtype", serializeDtype(v->dtype())},
+      });
+}
```
Any reason not to use the DTYPE macro for these identical looking definitions?

eellison replied:
You need the manual cast from `at::Half`, if theres a good way of using macro for all except that lmk

nickgg replied:
There is!
https://github.com/pytorch/pytorch/blob/master/torch/csrc/jit/tensorexpr/ir.h#L334 has an example of a macro that hits all scalar types + Half, but not Bool. It would be easy to switch this up to include Bool but not Half.

eellison replied:
I tried to do this, and ran into some weird macro-related issues I had a hard time figuring it out. I can try again if you think it's important.

nickgg commented on torch/csrc/jit/tensorexpr/ir_serializer.cpp:
```diff
@@ -0,0 +1,557 @@
+  visitBinaryOp(v, "Mod", this);
+}
+
+void IRSerializer::visit(const Max* v) {
+  visitBinaryOp(v, "Max", this);
+}
+
+void IRSerializer::visit(const Min* v) {
+  visitBinaryOp(v, "Min", this);
+}
+
```
You're discarding the `propagate_nans` field here.

The overlap in the IR between Min/Max and MinTerm/MaxTerm is confusing, it's not obvious which one should be used here. We should (separately) fix the comments on the Term nodes.

nickgg commented on torch/csrc/jit/tensorexpr/ir_serializer.cpp:
```diff
@@ -0,0 +1,557 @@
+
+void IRSerializer::visit(const Block* v) {
+  JSON_CACHE_GUARD();
+  // TODO: why is Block using std::list instead of std::vector ?
```
Because we often prepend Stmts to Blocks, or splice them, etc. I think that's why, anyway.

That said, it should probably be a std::deque.

ZolotukhinM replied:
Yeah, correct. `std::list` is better than `std::vector` (and also than `std::deque`) when we have operations like `insertBefore` and `insertAfter`.

nickgg replied:
As an aside @ZolotukhinM, why is list better than deque in that instance? The overhead of changing both sides of a doubly linked list doesn't seem significant at the N we're likely to see here. Or is there something about the implementation of deque I'm not aware of?

ZolotukhinM replied:
The difference is not about inserting an item to the beginning or the end of the list - it's about inserting into the middle. I thought `deque` didn't allow to do that easly? Am I mistaken?

nickgg replied:
@ZolotukhinM I found this looking around https://www.codeproject.com/Articles/5425/An-In-Depth-Study-of-the-STL-Deque-Container which was helpful in understanding deque.

From my reading, `deque` is a nice wrapper around `list<vector>` - gives you a midpoint implementation which supports random access but doesn't punish insertion as badly. I'd be curious about which is better for our purposes here - we don't actually allow random access of the stmt list, only iteration. Would it be helpful to find the nth stmt?

ZolotukhinM replied:
> From my reading, deque is a nice wrapper around list<vector>
What do you mean by `list<vector>`? Did you mean `list<Stmt>`?

If I understood correctly, the article compares `deque` with `vector` - when compare these two the difference can be illustrated by a simple insertion (anywhere, except at the end). I.e. if we insert in the beginning of a block `deque` would be much better than `vector`, since complexity of the operation would be O(1) rather than O(N).

When we compare `deque` with `list`, the difference comes into play when we try to insert in the middle of the block. According to https://en.cppreference.com/w/cpp/container/deque/insert, complexity of insertion to the middle would be "Constant plus linear in the lesser of the distances between pos and either of the ends of the container." Do we really need to ever insert in the middle? Yes, we do - imagine we have a block with N loops and we split each of them with tail. During each of the splits, we need to insert the tail loop right after the given loop:
https://github.com/pytorch/pytorch/blob/822efb7275939902fab53ef578c0e6ef5f63e580/torch/csrc/jit/tensorexpr/loopnest.cpp#L1031
If we use `deque` (or `vector` fwiw), the complexity of this pass would be worse than if we use `list`.

Do I somehow read it wrong? Also, what is a benefit of using `deque` vs `list`?

bertmaher replied:
Otoh insertion into a vector of trivially movable types (like Stmt*) can be done with memmove, whereas insertion into a `list` will allocate (always, I think)... but now I'm just being a pain 😈

nickgg commented on torch/csrc/jit/tensorexpr/ir_serializer.cpp:
```diff
@@ -0,0 +1,557 @@
+
+void IRSerializer::visit(const Allocate* v) {
+  JSON_CACHE_GUARD();
+  // TODO: why is allocate a stmt and not an expr ?
```
Think of it as `int A[size];` not `malloc(size)`.

We could make it and Expr, but I'd want to restrict it to appearing only on the RHS of a `Let` statement, since I don't know of any reason we'd take the added complexity of handing allocation in the RHS of Stores, or worse internal For stopping conditions and the like. In that case I don't see the advantage of it being an Expr.

eellison replied:
Thanks, makes sense. yea most of these comments were just for my own understanding...

nickgg replied:
No worries, happy to answer questions in whichever format :)

nickgg commented on torch/csrc/jit/tensorexpr/ir_serializer.cpp:
```diff
@@ -0,0 +1,557 @@
+
+void IRSerializer::visit(const Term* v) {
+  JSON_CACHE_GUARD();
+  // TODO: why is Term not a a kernel scoped object ?
```
This seems like a bug, it should be cleaned up at scope exit. Ideally it should be cleaned up at the end of simplification, but I don't think we have hierarchical kernel scope arenas yet.

Term and Polynomial (and MaxTerm and MinTerm) are temporary nodes that should exist only during arithmetic simplification, so it doesn't seem important to include them here.

nickgg commented on torch/csrc/jit/tensorexpr/ir_serializer.cpp:
```diff
@@ -0,0 +1,557 @@
+void IRSerializer::visit(const ReduceOp* v) {
+  assert(false);
+  // can't serialize lambda reduction interaction
+  // TOOD: reduction interaction should be a Function Node or something else
```
👍

nickgg reviewed (changes requested):
Looks okay generally, but there are few bugs here that should be covered by tests.

eellison reviewed (commented):
@nickgg thanks for your comments! Sorry about the stray comments and commented out code... thought I deleted those but must not have re-pushed. 

I mainly did json bc that was the initial suggestion. Certainly a lisp-style would work as well. I do think theres the potential for less regular Nodes in the future, and json is a little more flexible.  

Agreed that this needs much more testing - once I figure out the Tensor serialization bug I plan on adding a bunch more. IMO it would be nice if the serialization test code gets interspersed with the rest of the testing code. Otherwise it's extremely easy to add a new Node and not add serialization for it.

I'm happy to take a look at updating the reductions, I would probably need a bit of guidance on it.

nickgg commented:
I'm curious about less regular nodes you foresee - the regularity of the IR is the big difference between the NNC IR and the JIT IR afaics, so my preference would definitely be to keep any node that cannot be lowered to our IR at the JIT level. If you have a counter example I'd love to start thinking about it earlier rather than later.

Happy to help guide on the reduction refactor, hit me up on chat if and when you're ready to go.

Commit:
Update on "[TensorExpr] Add serializer / deserializer"


This PR adds conversion to and from json to NNC. I have taken a dependency on https://github.com/nlohmann/json for json, which I talked to @malfet, the build lead about. 

A NNC Expression like the following:
`y * 2.f + (float(x) * 3.f + 4.f * y)`

Gets serialized as:
```
{
  "dtype": "Float",
  "expr_type": "Add",
  "json_id": 10,
  "lhs": {
    "dtype": "Float",
    "expr_type": "Mul",
    "json_id": 2,
    "lhs": {
      "dtype": "Float",
      "expr_type": "Var",
      "json_id": 0,
      "name": "y"
    },
    "rhs": {
      "dtype": "Float",
      "expr_type": "FloatImm",
      "json_id": 1,
      "value": 2
    }
  },
  "rhs": {
    "dtype": "Float",
    "expr_type": "Add",
    "json_id": 9,
    "lhs": {
      "dtype": "Float",
      "expr_type": "Mul",
      "json_id": 6,
      "lhs": {
        "dtype": "Float",
        "expr_type": "Cast",
        "json_id": 4,
        "src_value": {
          "dtype": "Half",
          "expr_type": "Var",
          "json_id": 3,
          "name": "x"
        }
      },
      "rhs": {
        "dtype": "Float",
        "expr_type": "FloatImm",
        "json_id": 5,
        "value": 3
      }
    },
    "rhs": {
      "dtype": "Float",
      "expr_type": "Mul",
      "json_id": 8,
      "lhs": {
        "dtype": "Float",
        "expr_type": "FloatImm",
        "json_id": 7,
        "value": 4
      },
      "rhs": {
        "dtype": "Float",
        "expr_type": "Var",
        "json_id": 0,
        "name": "y"
      }
    }
  }
```

A few notes about implementation: 

- JSON has no builtin way to resolve references, so serialized KernelScopedObjects are cached, and assigned a json_id. When we deserialize, we look if we already deserialized a node with that ID in order to maintain multiple references to an object. 

- This memoization is currently broken just for Tensors, so each repeated call to a Tensor increments the name once. I've tried debugging it and would appreciate some help here 😛 This may be because there is no visitor defined for Tensors/Functions, just FunctionCall nodes. I'm not sure exactly though...

- `ReduceOp` is not able to be serialized currently because `interaction` is stored as a lambda. I'm not sure why there is the indirection there instead of directly storing an Expression or function or something.

- Testing : this is certainly light on testing currently. In JIT, any time we compile a graph in test files we test that it serializes correctly. It would be nice to do this also for TensorExprs. This is currently blocked by ReduceOp and Tensor memoization issue I mentioned.





[ghstack-poisoned]

Commit:
Update on "[TensorExpr] Add serializer / deserializer"


This PR adds conversion to and from json to NNC. I have taken a dependency on https://github.com/nlohmann/json for json, which I talked to @malfet, the build lead about. 

A NNC Expression like the following:
`y * 2.f + (float(x) * 3.f + 4.f * y)`

Gets serialized as:
```
{
  "dtype": "Float",
  "expr_type": "Add",
  "json_id": 10,
  "lhs": {
    "dtype": "Float",
    "expr_type": "Mul",
    "json_id": 2,
    "lhs": {
      "dtype": "Float",
      "expr_type": "Var",
      "json_id": 0,
      "name": "y"
    },
    "rhs": {
      "dtype": "Float",
      "expr_type": "FloatImm",
      "json_id": 1,
      "value": 2
    }
  },
  "rhs": {
    "dtype": "Float",
    "expr_type": "Add",
    "json_id": 9,
    "lhs": {
      "dtype": "Float",
      "expr_type": "Mul",
      "json_id": 6,
      "lhs": {
        "dtype": "Float",
        "expr_type": "Cast",
        "json_id": 4,
        "src_value": {
          "dtype": "Half",
          "expr_type": "Var",
          "json_id": 3,
          "name": "x"
        }
      },
      "rhs": {
        "dtype": "Float",
        "expr_type": "FloatImm",
        "json_id": 5,
        "value": 3
      }
    },
    "rhs": {
      "dtype": "Float",
      "expr_type": "Mul",
      "json_id": 8,
      "lhs": {
        "dtype": "Float",
        "expr_type": "FloatImm",
        "json_id": 7,
        "value": 4
      },
      "rhs": {
        "dtype": "Float",
        "expr_type": "Var",
        "json_id": 0,
        "name": "y"
      }
    }
  }
```

A few notes about implementation: 

- JSON has no builtin way to resolve references, so serialized KernelScopedObjects are cached, and assigned a json_id. When we deserialize, we look if we already deserialized a node with that ID in order to maintain multiple references to an object. 

- This memoization is currently broken just for Tensors, so each repeated call to a Tensor increments the name once. I've tried debugging it and would appreciate some help here 😛 This may be because there is no visitor defined for Tensors/Functions, just FunctionCall nodes. I'm not sure exactly though...

- `ReduceOp` is not able to be serialized currently because `interaction` is stored as a lambda. I'm not sure why there is the indirection there instead of directly storing an Expression or function or something.

- Testing : this is certainly light on testing currently. In JIT, any time we compile a graph in test files we test that it serializes correctly. It would be nice to do this also for TensorExprs. This is currently blocked by ReduceOp and Tensor memoization issue I mentioned.





[ghstack-poisoned]

Commit:
Update on "[TensorExpr] Add serializer / deserializer"


This PR adds conversion to and from json to NNC. I have taken a dependency on https://github.com/nlohmann/json for json, which I talked to @malfet, the build lead about. 

A NNC Expression like the following:
`y * 2.f + (float(x) * 3.f + 4.f * y)`

Gets serialized as:
```
{
  "dtype": "Float",
  "expr_type": "Add",
  "json_id": 10,
  "lhs": {
    "dtype": "Float",
    "expr_type": "Mul",
    "json_id": 2,
    "lhs": {
      "dtype": "Float",
      "expr_type": "Var",
      "json_id": 0,
      "name": "y"
    },
    "rhs": {
      "dtype": "Float",
      "expr_type": "FloatImm",
      "json_id": 1,
      "value": 2
    }
  },
  "rhs": {
    "dtype": "Float",
    "expr_type": "Add",
    "json_id": 9,
    "lhs": {
      "dtype": "Float",
      "expr_type": "Mul",
      "json_id": 6,
      "lhs": {
        "dtype": "Float",
        "expr_type": "Cast",
        "json_id": 4,
        "src_value": {
          "dtype": "Half",
          "expr_type": "Var",
          "json_id": 3,
          "name": "x"
        }
      },
      "rhs": {
        "dtype": "Float",
        "expr_type": "FloatImm",
        "json_id": 5,
        "value": 3
      }
    },
    "rhs": {
      "dtype": "Float",
      "expr_type": "Mul",
      "json_id": 8,
      "lhs": {
        "dtype": "Float",
        "expr_type": "FloatImm",
        "json_id": 7,
        "value": 4
      },
      "rhs": {
        "dtype": "Float",
        "expr_type": "Var",
        "json_id": 0,
        "name": "y"
      }
    }
  }
```

A few notes about implementation: 

- JSON has no builtin way to resolve references, so serialized KernelScopedObjects are cached, and assigned a json_id. When we deserialize, we look if we already deserialized a node with that ID in order to maintain multiple references to an object. 

- This memoization is currently broken just for Tensors, so each repeated call to a Tensor increments the name once. I've tried debugging it and would appreciate some help here 😛 This may be because there is no visitor defined for Tensors/Functions, just FunctionCall nodes. I'm not sure exactly though...

- `ReduceOp` is not able to be serialized currently because `interaction` is stored as a lambda. I'm not sure why there is the indirection there instead of directly storing an Expression or function or something.

- Testing : this is certainly light on testing currently. In JIT, any time we compile a graph in test files we test that it serializes correctly. It would be nice to do this also for TensorExprs. This is currently blocked by ReduceOp and Tensor memoization issue I mentioned.





[ghstack-poisoned]

Commit:
Update on "[TensorExpr] Add serializer / deserializer"


This PR adds conversion to and from json to NNC. I have taken a dependency on https://github.com/nlohmann/json for json, which I talked to @malfet, the build lead about. 

A NNC Expression like the following:
`y * 2.f + (float(x) * 3.f + 4.f * y)`

Gets serialized as:
```
{
  "dtype": "Float",
  "expr_type": "Add",
  "json_id": 10,
  "lhs": {
    "dtype": "Float",
    "expr_type": "Mul",
    "json_id": 2,
    "lhs": {
      "dtype": "Float",
      "expr_type": "Var",
      "json_id": 0,
      "name": "y"
    },
    "rhs": {
      "dtype": "Float",
      "expr_type": "FloatImm",
      "json_id": 1,
      "value": 2
    }
  },
  "rhs": {
    "dtype": "Float",
    "expr_type": "Add",
    "json_id": 9,
    "lhs": {
      "dtype": "Float",
      "expr_type": "Mul",
      "json_id": 6,
      "lhs": {
        "dtype": "Float",
        "expr_type": "Cast",
        "json_id": 4,
        "src_value": {
          "dtype": "Half",
          "expr_type": "Var",
          "json_id": 3,
          "name": "x"
        }
      },
      "rhs": {
        "dtype": "Float",
        "expr_type": "FloatImm",
        "json_id": 5,
        "value": 3
      }
    },
    "rhs": {
      "dtype": "Float",
      "expr_type": "Mul",
      "json_id": 8,
      "lhs": {
        "dtype": "Float",
        "expr_type": "FloatImm",
        "json_id": 7,
        "value": 4
      },
      "rhs": {
        "dtype": "Float",
        "expr_type": "Var",
        "json_id": 0,
        "name": "y"
      }
    }
  }
```

A few notes about implementation: 

- JSON has no builtin way to resolve references, so serialized KernelScopedObjects are cached, and assigned a json_id. When we deserialize, we look if we already deserialized a node with that ID in order to maintain multiple references to an object. 

- This memoization is currently broken just for Tensors, so each repeated call to a Tensor increments the name once. I've tried debugging it and would appreciate some help here 😛 This may be because there is no visitor defined for Tensors/Functions, just FunctionCall nodes. I'm not sure exactly though...

- `ReduceOp` is not able to be serialized currently because `interaction` is stored as a lambda. I'm not sure why there is the indirection there instead of directly storing an Expression or function or something.

- Testing : this is certainly light on testing currently. In JIT, any time we compile a graph in test files we test that it serializes correctly. It would be nice to do this also for TensorExprs. This is currently blocked by ReduceOp and Tensor memoization issue I mentioned.





[ghstack-poisoned]

Commit:
Update on "[TensorExpr] Add serializer / deserializer"


This PR adds conversion to and from json to NNC. I have taken a dependency on https://github.com/nlohmann/json for json, which I talked to @malfet, the build lead about. 

A NNC Expression like the following:
`y * 2.f + (float(x) * 3.f + 4.f * y)`

Gets serialized as:
```
{
  "dtype": "Float",
  "expr_type": "Add",
  "json_id": 10,
  "lhs": {
    "dtype": "Float",
    "expr_type": "Mul",
    "json_id": 2,
    "lhs": {
      "dtype": "Float",
      "expr_type": "Var",
      "json_id": 0,
      "name": "y"
    },
    "rhs": {
      "dtype": "Float",
      "expr_type": "FloatImm",
      "json_id": 1,
      "value": 2
    }
  },
  "rhs": {
    "dtype": "Float",
    "expr_type": "Add",
    "json_id": 9,
    "lhs": {
      "dtype": "Float",
      "expr_type": "Mul",
      "json_id": 6,
      "lhs": {
        "dtype": "Float",
        "expr_type": "Cast",
        "json_id": 4,
        "src_value": {
          "dtype": "Half",
          "expr_type": "Var",
          "json_id": 3,
          "name": "x"
        }
      },
      "rhs": {
        "dtype": "Float",
        "expr_type": "FloatImm",
        "json_id": 5,
        "value": 3
      }
    },
    "rhs": {
      "dtype": "Float",
      "expr_type": "Mul",
      "json_id": 8,
      "lhs": {
        "dtype": "Float",
        "expr_type": "FloatImm",
        "json_id": 7,
        "value": 4
      },
      "rhs": {
        "dtype": "Float",
        "expr_type": "Var",
        "json_id": 0,
        "name": "y"
      }
    }
  }
```

A few notes about implementation: 

- JSON has no builtin way to resolve references, so serialized KernelScopedObjects are cached, and assigned a json_id. When we deserialize, we look if we already deserialized a node with that ID in order to maintain multiple references to an object. 

- This memoization is currently broken just for Tensors, so each repeated call to a Tensor increments the name once. I've tried debugging it and would appreciate some help here 😛 This may be because there is no visitor defined for Tensors/Functions, just FunctionCall nodes. I'm not sure exactly though...

- `ReduceOp` is not able to be serialized currently because `interaction` is stored as a lambda. I'm not sure why there is the indirection there instead of directly storing an Expression or function or something.

- Testing : this is certainly light on testing currently. In JIT, any time we compile a graph in test files we test that it serializes correctly. It would be nice to do this also for TensorExprs. This is currently blocked by ReduceOp and Tensor memoization issue I mentioned.





[ghstack-poisoned]

Commit:
Update on "[TensorExpr] Add serializer / deserializer"


This PR adds conversion to and from json to NNC. I have taken a dependency on https://github.com/nlohmann/json for json, which I talked to @malfet, the build lead about. 

A NNC Expression like the following:
`y * 2.f + (float(x) * 3.f + 4.f * y)`

Gets serialized as:
```
{
  "dtype": "Float",
  "expr_type": "Add",
  "json_id": 10,
  "lhs": {
    "dtype": "Float",
    "expr_type": "Mul",
    "json_id": 2,
    "lhs": {
      "dtype": "Float",
      "expr_type": "Var",
      "json_id": 0,
      "name": "y"
    },
    "rhs": {
      "dtype": "Float",
      "expr_type": "FloatImm",
      "json_id": 1,
      "value": 2
    }
  },
  "rhs": {
    "dtype": "Float",
    "expr_type": "Add",
    "json_id": 9,
    "lhs": {
      "dtype": "Float",
      "expr_type": "Mul",
      "json_id": 6,
      "lhs": {
        "dtype": "Float",
        "expr_type": "Cast",
        "json_id": 4,
        "src_value": {
          "dtype": "Half",
          "expr_type": "Var",
          "json_id": 3,
          "name": "x"
        }
      },
      "rhs": {
        "dtype": "Float",
        "expr_type": "FloatImm",
        "json_id": 5,
        "value": 3
      }
    },
    "rhs": {
      "dtype": "Float",
      "expr_type": "Mul",
      "json_id": 8,
      "lhs": {
        "dtype": "Float",
        "expr_type": "FloatImm",
        "json_id": 7,
        "value": 4
      },
      "rhs": {
        "dtype": "Float",
        "expr_type": "Var",
        "json_id": 0,
        "name": "y"
      }
    }
  }
```

A few notes about implementation: 

- JSON has no builtin way to resolve references, so serialized KernelScopedObjects are cached, and assigned a json_id. When we deserialize, we look if we already deserialized a node with that ID in order to maintain multiple references to an object. 

- This memoization is currently broken just for Tensors, so each repeated call to a Tensor increments the name once. I've tried debugging it and would appreciate some help here 😛 This may be because there is no visitor defined for Tensors/Functions, just FunctionCall nodes. I'm not sure exactly though...

- `ReduceOp` is not able to be serialized currently because `interaction` is stored as a lambda. I'm not sure why there is the indirection there instead of directly storing an Expression or function or something.

- Testing : this is certainly light on testing currently. In JIT, any time we compile a graph in test files we test that it serializes correctly. It would be nice to do this also for TensorExprs. This is currently blocked by ReduceOp and Tensor memoization issue I mentioned.





[ghstack-poisoned]
